lightning/src/chain/mod.rs

```diff
@@ -189,66 +189,79 @@ pub trait Confirm {
-	/// the channel which would invalidate previous ChannelMonitors are not made when a channel has
-	/// been "frozen".
-	///
-	/// Note that even if updates made after TemporaryFailure succeed you must still provide a
-	/// [`MonitorEvent::UpdateCompleted`] to ensure you have the latest monitor and re-enable
-	/// normal channel operation. Note that this is normally generated through a call to
-	/// [`ChainMonitor::channel_monitor_updated`].
-	///
-	/// Note that the update being processed here will not be replayed for you when you return a
-	/// [`MonitorEvent::UpdateCompleted`] event via [`Watch::release_pending_monitor_events`], so
-	/// you must store the update itself on your own local disk prior to returning a
-	/// TemporaryFailure. You may, of course, employ a journaling approach, storing only the
-	/// ChannelMonitorUpdate on disk without updating the monitor itself, replaying the journal at
-	/// reload-time.
+	/// submitting new commitment transactions to the counterparty. Once the update(s) which failed
+	/// have been successfully applied, a [`MonitorEvent::UpdateCompleted`] can be used to restore
+	/// the channel to an operational state.
+	///
+	/// Note that a given [`ChannelManager`] will *never* re-generate a [`ChannelMonitorUpdate`].
+	/// If you return this error you must ensure that it is written to disk safely before writing
+	/// the latest [`ChannelManager`] state, or you should return [`PermanentFailure`] instead.
+	///
+	/// Even when a channel has been "frozen", updates to the [`ChannelMonitor`] can continue to
+	/// occur (e.g. if an inbound HTLC which we forwarded was claimed upstream, resulting in us
+	/// attempting to claim it on this channel) and those updates must still be persisted.
+	///
+	/// No updates to the channel will be made which could invalidate other [`ChannelMonitor`]s
+	/// until a [`MonitorEvent::UpdateCompleted`] is provided, even if you return no error on a
+	/// later monitor update for the same channel.
 	///
 	/// For deployments where a copy of ChannelMonitors and other local state are backed up in a
 	/// remote location (with local copies persisted immediately), it is anticipated that all
-	/// updates will return TemporaryFailure until the remote copies could be updated.
+	/// updates will return [`UpdateInProgress`] until the remote copies could be updated.
 	///
-	/// [`ChainMonitor::channel_monitor_updated`]: chainmonitor::ChainMonitor::channel_monitor_updated
-	TemporaryFailure,
-	/// Used to indicate no further channel monitor updates will be allowed (eg we've moved on to a
-	/// different watchtower and cannot update with all watchtowers that were previously informed
-	/// of this channel).
+	/// [`PermanentFailure`]: ChannelMonitorUpdateResult::PermanentFailure
+	/// [`UpdateInProgress`]: ChannelMonitorUpdateResult::UpdateInProgress
+	/// [`ChannelManager`]: crate::ln::channelmanager::ChannelManager
+	UpdateInProgress,
```

Comment:
I think you might note the trade-off brought by `UpdateInProgress`, if you allow the channel manager to keep moving on its own without state being fully persisted you gain in latency at the HTLC relay-level. However, you're increasing your risks w.r.t to state replication. This trade-off might be acceptable for some routing hops node operators if in the future scoring algorithms starts to encompass latency.